ruby/ql/lib/codeql/ruby/ast/Method.qll

```diff
@@ -132,13 +139,7 @@ class Method extends MethodBase, TMethod {
    * ```
    */
   override predicate isPrivate() {
-    this = any(Private p).getMethod()
-    or
-    exists(Namespace c, Private p, int i, int j |
-      p.hasNoArg(c, i) and
-      this = c.getStmt(j) and
-      j > i
-    )
+    any(Private p).modifiesMethod(this.getEnclosingModule(), this.getName())
```

Comment:
This can still join on just `this.getName()`, so it is better to add a top-level predicate
```ql
pragma[noinline]
private predicate isDeclaredIn(MethodBase m, Namespace n, string name) {
  n = m.getEnclosingModule() and
  name = m.getName()
}
```

and then change the above to

```ql
exists(Namespace n, string name |
  any(Private p).modifiesMethod(n, name) and
  isDeclaredIn(this, n, name)
)
```

Reply:
Interesting! That seems to work, and from some rough testing it also looks faster as well. But I don't really understand why it is faster. Why is it that the previous version can join on just `this.getName()`, rather than `this.getName()` _and_ `this.getEnclosingModule()`? Are there some docs that I could read to better understand this?

Reply:
The code above involves three relations (`modifiesMethod`, `getEnclosingModule`, and `getName`), so the join-orderer has to pick one of the 6 possible join-orders. The join-order `modifiesMethod <-> getName <-> getEnclosingModule` is the one we want to avoid. By folding `getEnclosingModule` and `getName` into a separate predicate, we are effectively making sure that `modifiesMethod` joins on the two relations simultaneously, and there is only one possible (good) join-order.

You can try to look at the Datalog Intermediate Language (DIL) code before and after (right-click a query run in VS Code and select "View DIL").

Reply:
Ah that makes sense! Had I thought about it a bit harder I probably should have guessed something like that 🤔 . Thanks for the explanation.